### PR TITLE
Added an example transformation to the compiler extension sample

### PR DESCRIPTION
**Pull request by bettinaheim**

### CODE REVIEW

File: examples/CompilerExtensions/CustomExtension/Optimize.cs [hunk 1]
```diff
@@ -61,14 +62,16 @@ public bool Transformation(QsCompilation compilation, out QsCompilation transfor
...
+            static TransformationBase[] Script(ImmutableDictionary<QsQualifiedName, QsCallable> callables) => 
+                new TransformationBase[]
+                {
+                    new ConstantPropagation(callables),
+                    new VariableRemoval(),
+                    new StatementRemoval(true),
+                };
 
-            transformed = PreEvaluation.WithScript(Script, compilation);
+            transformed = new ListIdentifiers().OnCompilation(compilation); // adds a comment with all used identifers to each callable
```
Comment (ScottCarda-MS) [inline examples/CompilerExtensions/CustomExtension/Optimize.cs]: So this demo transformation is run as part of the PreEvaluation rewrite step. Is that correct?

File: examples/CompilerExtensions/CustomExtension/ListIdentifiers.cs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+                var cannotSet = this.CurrentCallable == null || id == null;
+                var exists = this.Identifiers.TryGetValue(this.CurrentCallable, out var ids);
+                if (cannotSet || !exists) return false;
+                else ids.Add(id);
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Helper class for ListIdentifiers that defines expression kind transformations.
```
[the remark this answers is not in the document]
Comment (ScottCarda-MS) [reply examples/CompilerExtensions/CustomExtension/ListIdentifiers.cs]: I think sticking to 'subtransformation' would be clearer.

[the remark this answers is not in the document]
Comment (bettinaheim) [reply examples/CompilerExtensions/CustomExtension/ListIdentifiers.cs]: rephrased

File: examples/CompilerExtensions/CustomExtension/ListIdentifiers.cs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                return base.OnIdentifier(sym, tArgs);
+            }
+        }
+
+        /// <summary>
+        /// Helper class for ListIdentifiers that defines namespace transformations. 
+        /// It adds a comment to each callable listing all identifiers used within the callable 
+        /// according to the information accumulated in the shared transformation state. 
+        /// </summary>
+        private class NamespaceTransformation
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply examples/CompilerExtensions/CustomExtension/ListIdentifiers.cs]: Yes, I changed the order.

File: examples/CompilerExtensions/CustomExtension/Optimize.cs (hunk 1 above)
Comment (bettinaheim) [reply examples/CompilerExtensions/CustomExtension/Optimize.cs]: That's correct.

File: examples/CompilerExtensions/CustomExtension/ListIdentifiers.cs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            : base(parent, TransformationOptions.NoRebuild)
+            { }
+
+            public override QsExpressionKind<TypedExpression, Identifier, ResolvedType> OnIdentifier
+                (Identifier sym, QsNullable<ImmutableArray<ResolvedType>> tArgs)
+            {
+                var name =
+                    sym is Identifier.LocalVariable var ? var.Item.Value :
+                    sym is Identifier.GlobalCallable global ? global.Item.ToString() :
+                    null;
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply examples/CompilerExtensions/CustomExtension/ListIdentifiers.cs]: I don't quite see why that would be cleaner. I understand preferences, but if-else expressions are pretty neat. :)

File: examples/CompilerExtensions/CustomExtension/ListIdentifiers.cs [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                var cannotSet = this.CurrentCallable == null || id == null;
+                var exists = this.Identifiers.TryGetValue(this.CurrentCallable, out var ids);
+                if (cannotSet || !exists) return false;
+                else ids.Add(id);
+                return true;
+            }
+        }
+
+        /// <summary>
+        /// Helper class for ListIdentifiers that defines namespace transformations.
```
Comment (ScottCarda-MS) [inline examples/CompilerExtensions/CustomExtension/ListIdentifiers.cs]: This is another use of 'helper class' that could maybe be rephrased.

Comment (bettinaheim) [reply examples/CompilerExtensions/CustomExtension/ListIdentifiers.cs]: done

**Pull request merged**